File: platform/paper/src/main/java/io/github/bindglam/neko/manager/ResourcePackManagerImpl.java
package io.github.bindglam.neko.manager;

import io.github.bindglam.neko.NekoPaperPlugin;
import io.github.bindglam.neko.content.feature.event.ResourcePackGenerationEvent;
import io.github.bindglam.neko.event.AsyncResourcePackGenerationEvent;
import io.github.bindglam.neko.utils.Constants;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.jetbrains.annotations.NotNull;
import team.unnamed.creative.ResourcePack;
import team.unnamed.creative.metadata.pack.PackFormat;
import team.unnamed.creative.metadata.pack.PackMeta;
import team.unnamed.creative.serialize.minecraft.MinecraftResourcePackWriter;

import java.io.File;
import java.util.concurrent.CompletableFuture;
import java.util.logging.Logger;

public final class ResourcePackManagerImpl implements ResourcePackManager, Managerial<NekoPaperPlugin>, Reloadable<NekoPaperPlugin> {
    private static final Logger LOGGER = Logger.getLogger(ResourcePackManagerImpl.class.getName());
    private static final File GENERATED_PACK_FILE = new File(Constants.DATA_FOLDER, "generated.zip");

    @Override
    public void preload(@NotNull Context<NekoPaperPlugin> context) {
    }

    @Override
    public void start(@NotNull Context<NekoPaperPlugin> context) {
        generateResourcePack();
    }

    @Override
    public void end(@NotNull Context<NekoPaperPlugin> context) {
    }

    @Override
    public @NotNull CompletableFuture<Void> generateResourcePack() {
        return CompletableFuture.runAsync(() -> {
            LOGGER.info("Generating resource pack...");

            ResourcePack resourcePack = ResourcePack.resourcePack();
            resourcePack.packMeta(PackMeta.of(
                    PackFormat.format(99, 1, Integer.MAX_VALUE),
                    Component.text("Created by Neko")
            ));

            RegistryManager.GlobalRegistries.registries().item().forEach(content ->
                    content.featureEventBus().call(new ResourcePackGenerationEvent(resourcePack)));

            Bukkit.getPluginManager().callEvent(new AsyncResourcePackGenerationEvent(resourcePack));

            MinecraftResourcePackWriter.minecraft().writeToZipFile(GENERATED_PACK_FILE, resourcePack);

            LOGGER.info("Successfully generated resource pack!");
        });
    }
}

File: core/src/main/kotlin/io/github/bindglam/neko/manager/ResourcePackManagerImpl.kt
package io.github.bindglam.neko.manager

import io.github.bindglam.neko.content.feature.FeatureContext
import io.github.bindglam.neko.event.AsyncResourcePackGenerationEvent
import io.github.bindglam.neko.utils.DATA_FOLDER
import io.github.bindglam.neko.utils.logger
import net.kyori.adventure.text.Component
import team.unnamed.creative.ResourcePack
import team.unnamed.creative.metadata.pack.PackFormat
import team.unnamed.creative.metadata.pack.PackMeta
import team.unnamed.creative.serialize.minecraft.MinecraftResourcePackWriter
import java.io.File
import java.util.concurrent.CompletableFuture

object ResourcePackManagerImpl : ResourcePackManager, Managerial, Reloadable {
    private val generatedPackFile = File(DATA_FOLDER, "generated.zip")

    override fun preload(context: Context) {
    }

    override fun start(context: Context) {
        generateResourcePack()
    }

    override fun end(context: Context) {
    }

    override fun generateResourcePack(): CompletableFuture<Void> = CompletableFuture.runAsync {
        logger().info("Generating resource pack...")

        val resourcePack = ResourcePack.resourcePack().also {
            it.packMeta(PackMeta.of(PackFormat.format(99, 1, Int.MAX_VALUE), Component.text("Created by Neko")))
        }

        RegistryManager.GlobalRegistries.registries().item().forEach { content -> content.features().forEach { it.pack(FeatureContext.Pack(content, resourcePack)) } }

        AsyncResourcePackGenerationEvent(resourcePack).callEvent()

        MinecraftResourcePackWriter.minecraft().writeToZipFile(generatedPackFile, resourcePack)

        logger().info("Successfully generated resource pack!")
    }
}